sequencer.c

```diff
@@ -2653,6 +2654,29 @@ int sequencer_rollback(struct repository *r, struct replay_opts *opts)
 	return -1;
 }
 
+int sequencer_skip(struct repository *r, struct replay_opts *opts)
+{
+	switch (opts->action) {
+	case REPLAY_REVERT:
+		if (!file_exists(git_path_revert_head(r)))
+			return error(_("no revert in progress"));
+		break;
+	case REPLAY_PICK:
+		if (!file_exists(git_path_cherry_pick_head(r)))
+			return error(_("no cherry-pick in progress"));
+		break;
+	default:
+		BUG("the control must not reach here.");
+	}
+
+	if (file_exists(git_path_seq_dir())) {
+		int code = rollback_single_pick(r);
+
+		return code ? code : sequencer_continue(r, opts);
+	} else
+		return rollback_single_pick(r);
+}
```

Comment:
This might be more consistent with Git's coding style if it was written like this:

```c
if (rollback_single_pick(r) ||
    (dir_exists(git_path_seq_dir()) && sequencer_continue(r, opts)))
        return error(_("could not skip cherry-pick"));
else
    return 0;

Reply:
Thanks, nice catch!

Reply:
@dscho wrote:
>if (rollback_single_pick(r) ||
&nbsp;&nbsp;&nbsp;&nbsp;(dir_exists(git_path_seq_dir() && sequencer_continue(r, opts)))
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return error(_("could not skip cherry-pick"));
else
&nbsp;&nbsp;&nbsp;&nbsp;return 0;

But this is not the exact solution since we want to run `sequencer_continue` if `dir_exists` only if `rollback_single_pick` ran successfully. But your suggestion introduces a bug because of "short circuit". I get the idea anyway. Thanks!

---

It made me remember these line from B. Stroustrup's [Book](https://www.amazon.com/Tour-C-Bjarne-Stroustrup/dp/9332555338):
**3.5.3 Error-Handling Alternatives**
> [...]
We throw an exception when:
&nbsp;&nbsp;&nbsp;&nbsp; * An error is so rare that a programmer is likely to forget to check for it. For example, when did you last check the return value of printf()?
[...]

Reply:
@r1walz wrote:
>  [...] But your suggestion introduces a bug because of "short circuit". I get the idea anyway. Thanks!

But I guess git will throw an error anyway while `sequencer_continue` runs without `rollback_single_pick` being successful. So there is no "bug". @dscho @tgummerer @newren what do you think?

Reply:
@dscho wrote:
>if (rollback_single_pick(r) ||
&nbsp;&nbsp;&nbsp;&nbsp;(dir_exists(git_path_seq_dir() && sequencer_continue(r, opts)))
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return error(_("could not skip cherry-pick"));
else
&nbsp;&nbsp;&nbsp;&nbsp;return 0;

But again, `sequencer_continue` returns non-zero exit codes in case of merge conflicts. We can't simply add `!` in front of it since then we will be printing `error: could not skip cherry-pick` for a "successful" cherry-picking.

Reply:
Returning a non-zero value is correct in that case though, right?  So I think just returning 1, instead of `error(...)` would be the right thing to do.  In fact that is what the code in this PR is doing currently (it might return a different number than 1, but without checking the code, I'm guessing we don't depend on the actual value of the return code from this function)

Reply:
How about something like:
```c
if (rollback_single_pick(r))
	return error(_("could not skip cherry-pick"));

return  file_exists(git_path_seq_dir()) && sequencer_continue(r, opts);
```

But again, then we are returning "could not skip cherry-pick" for when `rollback_single_pick` fails and
```
error: no cherry-pick or revert in progress
fatal: cherry-pick failed
```
if either `file_exists` or `sequencer_continue` **fails**. But it looks _structurally_ good.

Reply:
I'm not sure I understand what the argument against just returning one instead of `error()` in @dscho's original proposed rewrite here is?  If you want the additional message if `rollback_single_skip` fails, then what you propose above is good, but I don't see the need at this point.

Reply:
Maybe `sequencer_continue()` returns negative in case of unrecoverable error, positive in case of merge conflicts, and zero in case of success? If so, we cannot use:

> ```c
> return  file_exists(git_path_seq_dir()) && sequencer_continue(r, opts);
> ```

That would work in Javascript, but not in C, as the `&&` makes this a Boolean expression that does not, in fact, return the return value of `sequencer_continue()`. In C, we have to write:

```c
if (!file_exists(git_path_seq_dir())
    return 0;
return sequencer_continue(r, opts);
```

And yes, I like handling the rollback in a separate `if` statement, first thing:

```c
if (rollback_single_pick(r))
	return error(_("could not skip cherry-pick"));
if (!file_exists(git_path_seq_dir())
    return 0;
return sequencer_continue(r, opts);
```